apps/heft/src/plugins/NodeServicePlugin.ts

```diff
@@ -0,0 +1,343 @@
+            this._configuration.waitForTerminateMs = this._nodeServiceConfiguration.waitForTerminateMs;
+          }
+          if (this._nodeServiceConfiguration.waitForKillMs !== undefined) {
+            this._configuration.waitForKillMs = this._nodeServiceConfiguration.waitForKillMs;
+          }
+
+          this._shellCommand = (heftConfiguration.projectPackageJson.scripts || {})[
+            this._configuration.commandName
+          ];
+
+          if (this._shellCommand === undefined) {
+            if (this._configuration.ignoreMissingScript) {
+              this._logger.terminal.writeLine(
+                `The plugin is disabled because the project's package.json` +
+                  ` does not have a "${this._configuration.commandName}" script`
+              );
+            } else {
+              this._logger.terminal.writeErrorLine(
+                `The project's package.json does not have a "${this._configuration.commandName}" script`
+              );
+            }
+            this._pluginEnabled = false;
+          }
+        } else {
+          this._logger.terminal.writeVerboseLine(
+            'The plugin is disabled because its config file was not found: ' +
+              CoreConfigFiles.nodeServiceConfigurationLoader.projectRelativeFilePath
+          );
+        }
+      });
+
+      if (this._pluginEnabled) {
+        build.hooks.postBuild.tap(PLUGIN_NAME, (bundle: IPostBuildSubstage) => {
+          bundle.hooks.run.tapPromise(PLUGIN_NAME, async () => {
+            await this._runCommandAsync(heftSession, heftConfiguration);
+          });
+        });
+
+        build.hooks.compile.tap(PLUGIN_NAME, (compile: ICompileSubstage) => {
+          compile.hooks.afterEachIteration.tap(PLUGIN_NAME, this._compileHooks_afterEachIteration);
+        });
+      }
+    });
+  }
+
+  private async _runCommandAsync(
+    heftSession: HeftSession,
+    heftConfiguration: HeftConfiguration
+  ): Promise<void> {
+    this._logger.terminal.writeLine(`Starting Node service...`);
+
+    this._restartChild();
+  }
+
+  private _compileHooks_afterEachIteration = (): void => {
+    try {
+      // We've recompiled, so try launching again
+      this._childProcessFailed = false;
+
+      if (this._state === State.Stopped) {
+        // If we are already stopped, then extend the timeout
+        this._scheduleRestart(this._configuration.waitBeforeRestartMs);
+      } else {
+        this._stopChild();
+      }
+    } catch (error) {
+      console.error('UNCAUGHT: ' + error.toString());
+    }
+  };
+
+  private _restartChild(): void {
+    if (this._state !== State.Stopped) {
+      throw new InternalError('Invalid state');
+    }
+
+    this._state = State.Running;
+    this._clearTimeout();
+
+    this._logger.terminal.writeLine('Invoking command: ' + JSON.stringify(this._shellCommand!));
+
+    this._activeChildProcess = child_process.spawn(this._shellCommand!, {
```

Comment:
Would we ever support multiple child processes?

Reply:
I don't think so.  I think we would encourage people to model them as separate Rush projects.

However if you did want to do that, your shell command could spawn multiple processes and it would be handled by our strategy of terminating the entire tree.